### ISSUE TITLE
SQL queries with columns that are blank strings cause invalid metadata warning

### ISSUE THREAD

**Issue by justin-yan** [metabase/metabase#5625]
- My Browser: Google Chrome | 59.0.3071.115 (Official Build) (64-bit)
- My OS: Ubuntu 16.04 LTS
- Datastore DB: Redshift
- Metabase 0.25
- Metabase is run as a JAR on an Ubuntu 14.04LTS server.
- Metabase DB: MySQL (RDS)

We have a question that was previously working for Metabase 0.24, but after upgrading to Metabase 0.25, we are now seeing an error.

I'm still kicking the tires on the query to figure out what's going on, but a little more color on what this error message means would be fairly useful.  I haven't yet been able to isolate what, within the query, is causing this particular problem, but will update when I do.  The error and stack trace as pulled from the logs are below:

```
Jul 27 22:59:52 WARN metabase.query-processor :: {:status :failed,
:class clojure.lang.ExceptionInfo,
:error "Valid array of results column metadata maps",
:stacktrace
["query_processor.middleware.results_metadata$fn__27687$results__GT_column_metadata__27692.invoke(results_metadata.clj:35)"
"query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__27725.invoke(results_metadata.clj:89)"
"query_processor.middleware.resolve$resolve_middleware$fn__24632.invoke(resolve.clj:329)"
"query_processor.middleware.add_dimension_projections$add_remapping$fn__24399.invoke(add_dimension_projections.clj:148)"
"query_processor.middleware.expand$expand_middleware$fn__26198.invoke(expand.clj:536)"
"query_processor.middleware.add_row_count_and_status$add_row_count_and_status$fn__24716.invoke(add_row_count_and_status.clj:14)"
"query_processor.middleware.driver_specific$process_query_in_context$fn__26324.invoke(driver_specific.clj:12)"
"query_processor.middleware.resolve_driver$resolve_driver$fn__27735.invoke(resolve_driver.clj:14)"
"query_processor.middleware.cache$run_query_and_save_results_if_successful_BANG_.invokeStatic(cache.clj:113)"
"query_processor.middleware.cache$run_query_and_save_results_if_successful_BANG_.invoke(cache.clj:111)"
"query_processor.middleware.cache$run_query_with_cache.invokeStatic(cache.clj:124)"
"query_processor.middleware.cache$run_query_with_cache.invoke(cache.clj:121)"
"query_processor.middleware.cache$maybe_return_cached_results$fn__25104.invoke(cache.clj:147)"
"query_processor.middleware.catch_exceptions$catch_exceptions$fn__26246.invoke(catch_exceptions.clj:58)"
"query_processor$process_query.invokeStatic(query_processor.clj:124)"
"query_processor$process_query.invoke(query_processor.clj:120)"
"query_processor$run_and_save_query_BANG_.invokeStatic(query_processor.clj:232)"
"query_processor$run_and_save_query_BANG_.invoke(query_processor.clj:227)"
"query_processor$fn__27769$process_query_and_save_execution_BANG___27774$fn__27775.invoke(query_processor.clj:270)"
"query_processor$fn__27769$process_query_and_save_execution_BANG___27774.invoke(query_processor.clj:256)"
"api.card$run_query_for_card.invokeStatic(card.clj:491)"
"api.card$run_query_for_card.doInvoke(card.clj:477)"
"api.card$fn__29930$fn__29933$fn__29934.invoke(card.clj:498)"
"api.card$fn__29930$fn__29933.invoke(card.clj:497)"
"api.common.internal$do_with_caught_api_exceptions.invokeStatic(internal.clj:227)"
"api.common.internal$do_with_caught_api_exceptions.invoke(internal.clj:222)"
"api.card$fn__29930.invokeStatic(card.clj:493)"
"api.card$fn__29930.invoke(card.clj:493)"
"middleware$enforce_authentication$fn__28960.invoke(middleware.clj:122)"
"api.routes$fn__40132.invokeStatic(routes.clj:56)"
"api.routes$fn__40132.invoke(routes.clj:56)"
"routes$fn__40781$fn__40782.doInvoke(routes.clj:75)"
"routes$fn__40781.invokeStatic(routes.clj:71)"
"routes$fn__40781.invoke(routes.clj:71)"
"middleware$log_api_call$fn__29059$fn__29061.invoke(middleware.clj:331)"
"middleware$log_api_call$fn__29059.invoke(middleware.clj:330)"
"middleware$add_security_headers$fn__29009.invoke(middleware.clj:245)"
"middleware$bind_current_user$fn__28964.invoke(middleware.clj:142)"
"middleware$maybe_set_site_url$fn__29013.invoke(middleware.clj:268)"],
```

**Comment by camsaul** [comment]
That warning message is popping up because the metadata coming back with your query is in an invalid format, which is pretty strange...

I'm not sure what's wrong with it but that error message shouldn't hurt anything.

If you open your browser console and look at the API response that comes back from running the query and look at `data.cols` it should have the metadata for the query. If you want to post that here I can look at it and see if anything looks off

**Comment by justin-yan** [comment]
Hi there!

Thank you for the quick response - I was just able to isolate the problem.  The query had an empty string `"   "` as a column name - once this was changed to an alphanum, the problem was resolved.  For example:

This works fine:
```
select transaction_key from transactions limit 1;
```

This query results in the above error:
```
select transaction_key as "    " from transactions limit 1;
```

If that isn't the expected behavior, then feel free to keep this open as a bug though we're OK for now since we can just rename our columns (FWIW, this did work in Metabase 0.24 though perhaps purely by coincidence).

Otherwise, feel free to close this! (Though if there's a way to make the error message a little bit more informative, that would help a lot)

**Comment by camsaul** [comment]
@justin-yan I think we might be being overly strict and assuming all columns come back with non-blank names. Thanks for figuring out that that was the root cause so quickly. We can keep this open and address it at some point in the future

**Comment by justin-yan** [comment]
My pleasure - thanks for the great product :-)

I generated a few other cases because I was curious:

```
select transaction_key as "<columnname>" from transactions limit 5;
```

Column names that don't work:

```
- // hyphen
_ // underscore
__ // two underscores
```

Almost every other special character I tried does seem to actually work, along with things like `_test`.  I suspect that might be because the "field rename prettifier" eliminates hyphens and underscores resulting in an empty string.

Hope that helps, and thanks again!
